python/ray/ray_operator/operator.py

```diff
@@ -181,162 +172,107 @@ def write_config(self) -> None:
-
-        # Update if there's been a change to the spec or if we're attempting
-        # recovery from autoscaler failure.
-        if spec_changed or ray_restart_required:
-            cluster_status_q.put((cluster_name, cluster_namespace,
-                                  STATUS_UPDATING))
-            ray_cluster.set_config(cluster_config)
-            # Trigger Ray restart only if there's been a failure.
-            ray_cluster.create_or_update(restart_ray=ray_restart_required)
-            cluster_status_q.put((cluster_name, cluster_namespace,
-                                  STATUS_RUNNING))
-
-    elif event_type == "DELETED":
-        ray_cluster = ray_clusters[cluster_identifier]
-        ray_cluster.clean_up()
-        del ray_clusters[cluster_identifier]
-
-
-def status_handling_loop():
-    while True:
-        cluster_name, cluster_namespace, phase = cluster_status_q.get()
+        phase = cluster_status_q.get_nowait()
         try:
-            operator_utils.set_status(cluster_name, cluster_namespace, phase)
+            operator_utils.set_status(name, namespace, phase)
         except Exception:
-            log_prefix = ",".join(cluster_name, cluster_namespace)
+            log_prefix = ",".join([name, namespace])
             logger.exception(f"{log_prefix}: Error setting RayCluster status.")
 
 
-def main() -> None:
-    # Run status-handling loop.
-    status_handler = threading.Thread(target=status_handling_loop, daemon=True)
-    status_handler.start()
-
-    # Make directory for Ray cluster configs
-    if not os.path.isdir(operator_utils.RAY_CONFIG_DIR):
-        os.mkdir(operator_utils.RAY_CONFIG_DIR)
+@kopf.on.resume('rayclusters')
+@kopf.on.create('rayclusters')
+def create_fn(body, name, namespace, logger, **kwargs):
+    cluster_config = operator_utils.cr_to_config(body)
+    cluster_identifier = (name, namespace)
+    log_prefix = ",".join(cluster_identifier)
 
-    while True:
-        # This outer loop waits for creation of a RayCluster CRD if it hasn't
-        # already been created.
-        try:
-            # Enter main event loop.
-            run_event_loop()
-        except ApiException as e:
-            if e.status == 404:
-                logger.warning("Waiting for creation of the RayCluster CRD")
-                time.sleep(5)
-            else:
-                logger.error("Failed to enter operator event loop.")
-                # Unforeseen startup error. Operator pod is
-                # likely to end up in a crash loop.
-                raise
-
-
-if __name__ == "__main__":
-    main()
+    operator_utils.check_redis_password_not_specified(
+        cluster_config, cluster_identifier)
+
+    ray_cluster = RayCluster(cluster_config)
+    ray_clusters[cluster_identifier] = ray_cluster
+    ray_cluster.cluster_status_q.put(STATUS_UPDATING)
+
+    # TODO if the Ray Cluster is healthy following a resume event, we should
+    #  not restart the Ray processes
```

Comment:
Need to squint at this a little more -- but I think we'd want to make sure that Ray processes don't restart with resume in this PR (as, they don't restart with operator restart currently)

"resume" means the operator is picking up after restarting, right?

Reply:
After double checking this, it should be fine actually. Because `restart_ray` is set to `False` by default, it will not restart the cluster unless the Ray head process is down.

https://github.com/ray-project/ray/blob/d107cca1aa5bd8ada960fd279296eef0a3b9bc7b/python/ray/ray_operator/operator.py#L97